jwst/cube_build/src/cube_match_internal.c

```diff
@@ -0,0 +1,484 @@
+#include <math.h>
+#include <stdio.h>
+#include <Python.h>
+#include <stdbool.h>
+#include <numpy/arrayobject.h>
+#include <numpy/npy_math.h>
+
+#define PY_ARRAY_UNIQUE_SYMBOL _jwst_cube_match_sky_numpy_api    //WHAT IS THIS AND WHERE IS IT USED???
+#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
+
+
+// routines used from cube_utils.c
+extern double sh_find_overlap(const double xcenter, const double ycenter, 
+		      const double xlength, const double ylength,
+			      double xPixelCorner[],double yPixelCorner[]);
+
+
+
+extern double find_area_quad(double MinX, double MinY, double Xcorner[], double Ycorner[]);
+
+
+//_______________________________________________________________________
+// Allocate the memory to for the spaxel arrays in the cube
+//_______________________________________________________________________
+
+int mem_alloc(int nelem, double **fluxv, double **weightv, double **varv,  double **ifluxv) {
+
+    double *f, *w, *v, *i;
+    const char *msg = "Couldn't allocate memory for output arrays.";
+    
+    // flux:
+    f = (double*)malloc(nelem* sizeof(double));
+    if (f) {
+        *fluxv = f;
+    } else {
+        PyErr_SetString(PyExc_MemoryError, msg);
+        return 1;
+    }
+
+    // weight:
+    w = (double*)malloc(nelem* sizeof(double));
+    if (w) {
+        *weightv = w;
+    } else {
+        PyErr_SetString(PyExc_MemoryError, msg);
+        return 1;
+    }
+
+    // variance:
+    v = (double*)malloc(nelem* sizeof(double));
+    if (v) {
+        *varv = v;
+    } else {
+        PyErr_SetString(PyExc_MemoryError, msg);
+        return 1;
+    }
+    // iflux
+    i = (double*)malloc(nelem* sizeof(double));
+    if (i) {
+        *ifluxv = i;
+    } else {
+        PyErr_SetString(PyExc_MemoryError, msg);
+        return 1;
+    }
+
+    return 0;
+}
+
+//_______________________________________________________________________
+//  based on a 2-D drizzling type of algorithm find the contribution of
+// each detector flux to the IFU cube
+//_______________________________________________________________________
+
+int match_detector_cube(int instrument, int naxis1, int naxis2, int nz, int npt, int ncube, int na,
+			double crval_along, double cdelt_along, double crval3, double cdelt3,
+			double *a1, double *a2,double *a3, double*a4,
+			double *lam1, double *lam2, double*lam3, double *lam4,
+			double *acoord, double *zcoord, int ss,
+			double *pixel_flux, double *pixel_err,
+			double **spaxel_flux, double **spaxel_weight, double **spaxel_var,double **spaxel_iflux){
+
+  double *fluxv = NULL, *weightv=NULL, *varv=NULL ;  // vectors for spaxel 
+  double *ifluxv = NULL;  // vector for spaxel
+
+  // allocate memory to hold output 
+  if (mem_alloc(ncube, &fluxv, &weightv, &varv, &ifluxv)) return 1;
+    
+  double set_zero=0.0;
+  // Set all data to zero
+  for (int i = 0; i < ncube; i++){
+    varv[i] = set_zero;
+    fluxv[i] = set_zero;
+    ifluxv[i] = set_zero;
+    weightv[i] = set_zero;
+  }
```

Comment:
I would suggest that 1) `mem_alloc` be renamed to something less generic and 2) use `calloc` in order to initialize arrays to 0. For example:
```suggestion
int alloc_flux_arrays(int nelem, double **fluxv, double **weightv, double **varv,  double **ifluxv) {
    const char *msg = "Couldn't allocate memory for output arrays.";
    
    // flux:
    if (!(*fluxv  = (double*)calloc(nelem, sizeof(double)))) {
        PyErr_SetString(PyExc_MemoryError, msg);
        goto failed_mem_alloc;
    }

    // weight:
    if (!(*weightv = (double*)calloc(nelem, sizeof(double)))) {
        PyErr_SetString(PyExc_MemoryError, msg);
        goto failed_mem_alloc;
    }

    // variance:
    if (!(*varv = (double*)calloc(nelem, sizeof(double)))) {
        PyErr_SetString(PyExc_MemoryError, msg);
        goto failed_mem_alloc;
    }
    // iflux
    if (!(*ifluxv = (double*)calloc(nelem, sizeof(double)))) {
        PyErr_SetString(PyExc_MemoryError, msg);
        goto failed_mem_alloc;
    }

    return 0;
    
failed_mem_alloc:
    free(*fluxv);
    free(*weightv);
    free(*varv);
    free(*ifluxv);
}

//_______________________________________________________________________
//  based on a 2-D drizzling type of algorithm find the contribution of
// each detector flux to the IFU cube
//_______________________________________________________________________

int match_detector_cube(int instrument, int naxis1, int naxis2, int nz, int npt, int ncube, int na,
			double crval_along, double cdelt_along, double crval3, double cdelt3,
			double *a1, double *a2,double *a3, double*a4,
			double *lam1, double *lam2, double*lam3, double *lam4,
			double *acoord, double *zcoord, int ss,
			double *pixel_flux, double *pixel_err,
			double **spaxel_flux, double **spaxel_weight, double **spaxel_var,double **spaxel_iflux){

  double *fluxv, *weightv, *varv;  // vectors for spaxel 
  double *ifluxv;  // vector for spaxel

  // allocate memory to hold output 
  if (alloc_flux_arrays(ncube, &fluxv, &weightv, &varv, &ifluxv)) return 1;
    
```